collections/nemo_nlp/nemo_nlp/modules/losses.py

```diff
@@ -184,19 +185,37 @@ def create_ports():
         }
         return input_ports, output_ports
 
-    def __init__(self, num_slots, **kwargs):
+    def __init__(self,
+                 num_slots,
+                 slot_classes_loss_weights=None,
+                 intent_classes_loss_weights=None,
+                 intent_loss_weight=0.6,
+                 **kwargs):
         LossNM.__init__(self, **kwargs)
         self.num_slots = num_slots
-        self._criterion = nn.CrossEntropyLoss()
+        self.intent_loss_weight = intent_loss_weight
+
+        # For weighted loss to tackle class imbalance
+        if len(slot_classes_loss_weights) > 0:
+            self.slot_classes_loss_weights = \
+                torch.FloatTensor(slot_classes_loss_weights).to(self._device)
+
+        if len(intent_classes_loss_weights) > 0:
+            self.intent_classes_loss_weights = \
+                torch.FloatTensor(intent_classes_loss_weights).to(self._device)
+
+        self._criterion_intent = nn.CrossEntropyLoss(
+            weight=self.intent_classes_loss_weights)
+        self._criterion_slot = nn.CrossEntropyLoss(
+            weight=self.slot_classes_loss_weights)
 
     def _loss_function(self,
                        intent_logits,
                        slot_logits,
                        loss_mask,
                        intents,
-                       slots,
-                       intent_loss_weight=0.6):
-        intent_loss = self._criterion(intent_logits, intents)
+                       slots):
+        intent_loss = self._criterion_intent(intent_logits, intents)
 
         active_loss = loss_mask.view(-1)
```

Comment:
loss_mask should be int or long now to conform to the new version of pytorch, so you active_loss needs to be calculated the following way:
`active_loss = loss_mask.view(-1) > 0.5`